src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs

```diff
@@ -0,0 +1,846 @@
+            /// A potential filter character is something that can filter a completion lists and is
+            /// *guaranteed* to not be a commit character.
+            /// </summary>
+            private static bool IsPotentialFilterCharacter(char c)
+            {
+                // TODO(cyrusn): Actually use the right Unicode categories here.
+                return char.IsLetter(c)
+                    || char.IsNumber(c)
+                    || c == '_';
+            }
+
+            private ItemSelection UpdateSelectionBasedOnSuggestedDefaults(IReadOnlyList<MatchResult<VSCompletionItem>> items, ItemSelection itemSelection)
+            {
+                // Editor doesn't provide us a list of "default" items.
+                if (_data.Defaults.IsDefaultOrEmpty)
+                    return itemSelection;
+
+                // "Preselect" is only used when we have high confidence with the selection, so don't override it.
+                var selectedItem = items[itemSelection.SelectedItemIndex].RoslynCompletionItem;
+                if (selectedItem.Rules.MatchPriority >= MatchPriority.Preselect)
+                    return itemSelection;
+
+                var tick = Environment.TickCount;
+
+                var useAggressiveDefaultsMatching = _session.TextView.Options.GetOptionValue<bool>(AggressiveDefaultsMatchingOptionName);
+
+                var finalSelection = useAggressiveDefaultsMatching
+                    ? GetAggressiveDefaultsMatch(items, itemSelection)
+                    : GetDefaultsMatch(items, itemSelection);
+
+                AsyncCompletionLogger.LogGetDefaultsMatchTicksDataPoint(Environment.TickCount - tick);
+                return finalSelection;
+            }
+
+            /// <summary>
+            /// With aggressive matching turned on, the default was use as long as what the user typed in the applicable to span
+            /// was a case-insensitive prefix of the item.
+            /// The aggressive mode exist to make it easier for IntelliCode to experiment with default matching algorithm, so this is
+            /// not the default matching behavior we'd use and subject to further adjustment.
+            /// </summary>
+            private ItemSelection GetAggressiveDefaultsMatch(IReadOnlyList<MatchResult<VSCompletionItem>> items, ItemSelection intialSelection)
+            {
+                foreach (var defaultText in _data.Defaults)
+                {
+                    for (var i = 0; i < items.Count; ++i)
+                    {
+                        // Currently in aggressive mode, the first item matches a default will be selected
+                        // as long as the filter text is its prefix (case-insensitive).
+                        // e.g. a default "TaskCanceledException" would be a match for filter text "task" and therefore to be selected,
+                        // even if the CompletionService returns item for a local variable "task" which is an exact match.
```

Comment:
=-o

Note: i'm not sure the value of this experiment.  This behavior (esp. the cast shown in teh comment) would never be allowable.

Reply:
Yea, I agreed. This is now removed from the PR and tracked by https://github.com/dotnet/roslyn/issues/58169

Reply:
FYI @dpugh